### PR TITLE
Bug hunting

### LINKED ISSUE #14: No Rule name in PR description if the Law did not change anything
**Describe the bug**

In case a Law did not have a Rule in its chain which changed the code, the PR description will still contain the Law name (if there were changes made by other laws).

**To Reproduce**

Steps to reproduce the behaviour :

1. Create two laws with the following conditions and register them:
1.1 The first law has a rule which should change nothing
1.2 The second law has a rule which changes something
2. Run Hammurabi
3. The PR will contain `Law` names without failed or passed rules - the other law will have the rule.

**Expected behavior**

Those `Law`s are not shown which has no rule which changed the code.

**Screenshots**

N/A

**System info**

 - OS: Independent
 - Hammurabi version: 0.2.0
 - Python version: 3.7+

**Additional context**

N/A

### LINKED ISSUE #38: Do not raise lookup error when replace can not find target
Do not raise lookup error when replace can not find target (the expected end result) for

* ``SectionRenamed``
* ``LineReplaced``
* ``YAMLKeyRenamed``
* ``OptionRenamed``
 
Logic should be:

1. If both the item which should be replaced and the new item exists, then **raise an error**
2. If only the new exists, then **do nothing**
3. If the old item not exists, then **raise an error**
4. If the only old item exists, then **do the replace**

### PR DESCRIPTION
**Pull request by gabor-boros**
**Reason for the change**

* Resolves #14  
* Resolves #38 

**Description**

* Simplify yaml key rename logic
* ``SectionRenamed`` not raises error if old section name is not represented but the new one
* ``OptionRenamed`` not raises error if old option name is not represented but the new one
* ``LineReplaced`` not raises error if old line is not represented but the new one
* Fixed a dictionary traversal issue regarding yaml file support
* Fixed "Filed Rules" formatting of PR description by removing ``\xa0`` character
* Fixed no Rule name in PR description if the Law did not change anything issue
* Fixed empty passed rules shown in PR description issue

**Code examples**

N/A

**Checklist**

- [x] Unit tests created/updated
- [x] Documentation extended/updated

**References**

N/A

### CODE REVIEW

**Comment by gabor-boros** [comment]
Travis still not reporting. The related build: https://travis-ci.org/github/gabor-boros/hammurabi/builds/667597316

File: hammurabi/mixins.py [hunk 1]
```diff
@@ -202,17 +249,14 @@ def generate_pull_request_body(self, pillar) -> str:
         ]
 
         for law in pillar.laws:
+            if not [rule for rule in law.rules if rule.made_changes]:
```
Comment (balazs-tothfalussy) [inline hammurabi/mixins.py]: in general I dislike negative conditions and `continue`, so I suggest:
```
if [rule for rule in law.rules if rule.made_changes]:
    body.append(f"\n### {law.name}")
    .....
```

File: hammurabi/rules/text.py [hunk 2]
```diff
@@ -310,9 +310,16 @@ def __get_target_match(self, lines: List[str]) -> List[str]:
...
 
         target_match = list(filter(self.target.match, lines))
+        text = list(filter(lambda l: l.strip() == self.text, lines))
 
-        if not any(target_match):
-            raise LookupError(f'No matching line for "{self.target}"')
+        if target_match and text:
+            raise LookupError(f'Both "{str(self.target)}" and "{self.text}" exists')
+
+        if text:
```
Comment (balazs-tothfalussy) [inline hammurabi/rules/text.py]: is it possible that `not target_match` is True here? According to the doc above, there should be an exception.
It is also hard to follow what should happen with the 3 ifs after each other (like why exactly is this the order, what if we need to add any code in this function) - is it possible to refactor them into a more self explanatory version?

File: hammurabi/mixins.py (hunk 1 above)
Comment (gabor-boros) [reply hammurabi/mixins.py]: Simplified the logic. Only those laws will run which has rules which made modifications or failed

File: hammurabi/rules/text.py (hunk 2 above)
Comment (gabor-boros) [reply hammurabi/rules/text.py]: > is it possible that not target_match is True here? According to the doc above, there should be an exception.

Yes it is, in the case if the given `text` (the line which will replace the original line) is not in the file yet and the `target_match` (the line which should be replaced) is not in the file either. In that case we can not replace the expected line and we have no line already replaced. This scenario should result in an exception since if some rule is depending on the replacement, they would fail.

**Review by balazs-tothfalussy — APPROVED** [review]
